docs/SPEC.md

```diff
@@ -1286,13 +1329,25 @@ Within a stream each table's group key value is unique.
 
 [IMPL#463](https://github.com/influxdata/flux/issues/463) Specify the primitive types that make up stream and table types
 
-### Missing values
+### Missing values (null)
+
+`null` is a predeclared identifier representing a missing or unknown value.
+`null` is the only value comprising the _null type_.
+
+Any non-boolean operator that operates on basic types, returns _null_ when at least one of its operands is _null_.
+
+Because boolean types are nullable, Flux implements ternary logic as a way of handling boolean operators with _null_ operands.
+In this case, a _null_ operand is interpreted as an unknown value and we have the following definitions:
 
-A record may be missing a value for a specific column.
-Missing values are represented with a special _null_ value.
-The _null_ value can be of any data type.
+* not _null_ = _null_
+* _null_ or false = _null_
+* _null_ or true = true
+* _null_ or _null_ = _null_
+* _null_ and false = false
+* _null_ and true = _null_
+* _null_ and _null_ = _null_
```

Comment:
It would be good to include in here the behavior of `if b then ... else ...` where `b` is the `null` value.

It would also be good to call out explicitly that `==` will return `null` if either side is `null`, and even if both sides are `null`.  This can be a bit surprising the first time its encountered.

Reply:
Ah yes I knew I was forgetting something.

Reply:
Just analyzing the thing: `not null = null` implies that `null` is both true and false, this implies that `if null then ... else ...` should result in an error? 🤔 What do you think?

It is weird to take a decision when a value has no value...

Reply:
this is a bit what I'd been worried about, that we'll see lots of code like   `if r.b != null and r.b then ... else ...`

Reply:
I'm a bit concerned by this section too. I think this conflates two different solutions for null.

1. SQL method: Null is unknown and is the same as NaN and matches with nothing.
2. Most common programming languages method: Null is a distinct value that is usually falsey.

I don't think you can mix these. In SQL, null can't be evaluated as a boolean so you don't end up with this weird problem. But SQL also has explicit null checks with `IS NULL` and `IS NOT NULL`. Flux acts more like a programming language and uses that paradigm than SQL. I like SQL's method so it pains me a bit to say this, but I think we need to drop the idea that anything with null is null and instead just steal whatever Ruby does since it is the most similar to what we want to do. Python is kind of similar, but they're different because Python consider the zero integer a falsey value.

(Falsey means that the expression evaluates to false in a boolean context, but it is not the same thing as being the false value)

For those unfamiliar, here are the differences:

1. Python considers `False`, `None`, `0`, and empty sequences as falsey values.
2. Ruby considers `false` and `nil` as falsey values.

I think the second is closer to what we're going with. I typed in the above to Ruby and here's what I got:

```
not nil = true
nil or false = false
nil or true = true
nil or nil = nil
nil and false = nil
nil and true = nil
nil and nil = nil
```

The `not` predicate always returns a boolean value. The conditional operators are different though. They return the value that it short circuited on so in `nil or false` it saw that `nil` was falsey so it moved onto the second value which was `false` and returned that as the false value.

This might be a bit hard for our type system to encapsulate though so I would modify the above so that conditionals always return a boolean value.

```
not nil = true
nil or false = false
nil or true = true
nil or nil = false
nil and false = false
nil and true = false
nil and nil = false
```

Reply:
I don't believe this solution conflates the two solutions you mention @jsternberg . This solution is basically the SQL method which I find the most intuitive. I find it incredibly easy and natural to think of nulls as unknown values. In that context, operating on an unknown (non-boolean) value will always produce something that is unknown.

The only case where this is not true is in the case of boolean logic. But I still find the above truth table very intuitive.

Reply:
I think I can live with that since the result for most of these ends up being the same. We also had a conversation in Slack where `null == null` needs to return false for SQL interop and I think that makes sense because `5 + null == 3 + null` should probably be false and it would be false under this proposal.

The only one I do want to talk about is `not null`. It seems very unintuitive to me for `not null` to be `null` which would evaluate as false. Is this another sql-ism that will just have to exist?

Reply:
SQL's `null` is not quite the same as `NaN`.  If you do `count(floatCol)` in SQL, you will get a count of the non-null values in the column, including the NaNs.  Moreover, in SQL, a predicate of the form `foo = 3.0` where `foo` is `null` will return `unknown` (boolean null).  On the other hand if `foo` is `NaN`, then the predicate will return `false`.

Reply:
@jsternberg yes it is. However I always go back to thinking about nulls as unknown. If you don't know if `x` is _true_ or _false_, then you also don't know if the negation of `x` is _true_ or _false_. And so if you want to compare things against `null`. You just have to use the built-in `exists()` function. Or maybe it should just be called `isNull()`.

Either way we don't have to describe a bunch of edge cases in the SPEC. It's a very concise description of how nulls behave, so that for those who find it unintuitive at first, they can easily reason about it because everything is consistent. There are no weird unexpected edge cases.

At least that's my opinion.

Reply:
We talked about this in Slack. If you create a new issue to discuss how the null behavior should work on comparisons, I'll approve the current proposal and we can reevaluate the correct behavior after implementing one of the two by seeing it work in practice.